core/init.go

```diff
@@ -146,76 +163,108 @@ func initOauth2ForService(config *tomlConfig, url string, c *http.Client) {
 			lc.Error(s)
 		}
 	}
+}
 
+func initAuthmethod(config *tomlConfig, url string, c *http.Client) {
+	lc.Info(fmt.Sprintf("selected auth method as %s.", config.KongAuth.Name))
+	if config.KongAuth.Name == "jwt" {
+		initJWTAuth(config, url, c)
+	} else if config.KongAuth.Name == "oauth2" {
+		initOAuth2(config, url, c)
+	}
 }
-func initJWTAuthForService(config *tomlConfig, url string, c *http.Client) {
-	jwtParams := &KongJWTPlugin{
-		Name: config.KongAuth.Name,
+
+func initAuthmethodForRoute(config *tomlConfig, url string, c *http.Client, service string) {
+	lc.Info(fmt.Sprintf("selected auth method as %s.", config.KongAuth.Name))
+	if config.KongAuth.Name == "jwt" {
+		initJWTAuthForRoute(config, url, c, service)
+	} else if config.KongAuth.Name == "oauth2" {
+		initOauth2ForService(config, url, c, service)
+	}
+}
+
+func initOauth2ForService(config *tomlConfig, url string, c *http.Client, service string) {
+	oauth2Params := &KongOAuth2Plugin{
+		Name:                    config.KongAuth.Name,
+		Scope:                   OAuth2Scopes,
+		MandatoryScope:          "true",
+		EnableClientCredentials: "true",
+		EnableGlobalCredentials: "true",
 	}
 
-	req, err := sling.New().Base(url).Post(PluginsPath).BodyForm(jwtParams).Request()
+	req, err := sling.New().Base(url + "services/" + service + "/").Post(PluginsPath).BodyForm(oauth2Params).Request()
 	resp, err := c.Do(req)
 	if err != nil {
-		s := fmt.Sprintf("Failed to set up jwt authentication.")
+		s := fmt.Sprintf("Failed to set up oauth2 authentication with error %s.", err.Error())
+		defer resp.Body.Close()
```

Comment:
if the error was not nil here we don't need to close the body - it shouldn't have been allocated.

Instead put the `defer resp.Body.Close()` in the else clause here